config.go

```diff
@@ -2,44 +2,261 @@ package main
+type SystemCollectorConfig struct {
+	CollectionDeadlineDuration time.Duration `yaml:"collection_deadline"`
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface
+func (s *SystemCollectorConfig) UnmarshalYAML(unmarshal func(any) error) error {
+	*s = DefaultSystemCollector
+	type plain SystemCollectorConfig
+
+	if err := unmarshal((*plain)(s)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// TelemetryCollectorConfig is a prober configuration.
+type TelemetryCollectorConfig struct {
+	CollectionDeadlineDuration time.Duration `yaml:"collection_deadline"`
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface
+func (t *TelemetryCollectorConfig) UnmarshalYAML(unmarshal func(any) error) error {
+	*t = DefaultTelemetryCollector
+	type plain TelemetryCollectorConfig
+
+	if err := unmarshal((*plain)(t)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Module is a struct which represents some particular behavior the redfish_exporter should have
+// when executed against a host.
+// Modules are expected to specify a 'prober', and then a particular collector.
+type Module struct {
+	Prober             string                   `yaml:"prober"`
+	GPUCollector       GPUCollectorConfig       `yaml:"gpu_collector"`
+	ChassisCollector   ChassisCollectorConfig   `yaml:"chassis_collector"`
+	ManagerCollector   ManagerCollectorConfig   `yaml:"manager_collector"`
+	SystemCollector    SystemCollectorConfig    `yaml:"system_collector"`
+	TelemetryCollector TelemetryCollectorConfig `yaml:"telemetry_collector"`
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface
+func (m *Module) UnmarshalYAML(unmarshal func(any) error) error {
+	*m = DefaultModule
+	type plain Module
+	if err := unmarshal((*plain)(m)); err != nil {
+		return err
+	}
+	if m.Prober == "" {
+		return fmt.Errorf("modules require a prober to be set")
+	}
+	return nil
+}
+
+// Collector creates and returns a prometheus.Collector from the Module, based on the Prober.
+// Both redfish client and logger are common dependencies for any redfish_exporter collector,
+// and must be provided as inputs.
+func (m *Module) Collector(rfClient *gofish.APIClient, logger *slog.Logger) prometheus.Collector {
+	switch m.Prober {
+	case "gpu_collector":
+		return collector.NewGPUCollector(rfClient, logger)
+	case "chassis_collector":
+		return collector.NewChassisCollector(rfClient, logger)
+	case "manager_collector":
+		return collector.NewManagerCollector(rfClient, logger)
+	case "system_collector":
+		return collector.NewSystemCollector(rfClient, logger)
+	case "telemetry_collector":
+		return collector.NewTelemetryCollector(rfClient, logger)
+	default:
+	}
+	return nil
+}
+
+// Config represents the redfish_exporter config file
 type Config struct {
 	Hosts    map[string]HostConfig `yaml:"hosts"`
 	Groups   map[string]HostConfig `yaml:"groups"`
 	Loglevel string                `yaml:"loglevel"`
+	Modules  map[string]Module     `yaml:"modules"`
```

Comment:
Can you update `config_example.yml` so that users know what/how modules are configured?